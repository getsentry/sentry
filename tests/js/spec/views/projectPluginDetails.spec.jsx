import React from 'react';

import {mountWithTheme} from 'sentry-test/enzyme';

import ProjectPluginDetailsContainer, {
  ProjectPluginDetails,
} from 'app/views/settings/projectPlugins/details';

jest.mock('jquery');

describe('ProjectPluginDetails', function() {
  let component;
  const routerContext = TestStubs.routerContext();
  const {organization, project} = routerContext.context;
  const org = organization;
  const plugins = TestStubs.Plugins();
  const plugin = TestStubs.Plugin();
  const pluginId = plugin.id;

  beforeAll(function() {
    jest.spyOn(console, 'info').mockImplementation(() => {});
  });

  beforeEach(function() {
    MockApiClient.addMockResponse({
      url: `/projects/${org.slug}/${project.slug}/plugins/`,
      method: 'GET',
      body: plugins,
    });

    MockApiClient.addMockResponse({
      url: `/projects/${org.slug}/${project.slug}/plugins/${pluginId}/`,
      method: 'DELETE',
    });

    MockApiClient.addMockResponse({
      url: `/projects/${org.slug}/${project.slug}/plugins/${pluginId}/`,
      method: 'GET',
      body: plugin,
    });

    MockApiClient.addMockResponse({
      url: `/projects/${org.slug}/${project.slug}/plugins/${pluginId}/`,
      method: 'POST',
      body: {
        ...plugin,
        config: [{value: 'default'}],
      },
    });

    component = mountWithTheme(
      <ProjectPluginDetailsContainer
        organization={org}
        project={project}
        params={{orgId: org.slug, projectId: project.slug, pluginId: 'amazon-sqs'}}
        location={TestStubs.location()}
      />,
      routerContext
    );
  });

  afterAll(function() {
    // eslint-disable-next-line no-console
    console.info.restore();
  });

  it('renders', function() {
    expect(component).toMatchSnapshot();
  });

  it('resets plugin', function() {
    // Test component instead of container so that we can access state
    const wrapper = mountWithTheme(
      <ProjectPluginDetails
        organization={org}
        project={project}
        plugins={TestStubs.Plugins()}
        params={{orgId: org.slug, projectId: project.slug, pluginId: 'amazon-sqs'}}
        location={TestStubs.location()}
      />,
      routerContext
    );

    const btn = wrapper.find('button').at(1);
    btn.simulate('click');
    expect(wrapper.state().pluginDetails.config[0].value).toBe('default');
  });

  it('enables/disables plugin', function(done) {
    const btn = component.find('button').first();
    expect(btn.text()).toBe('Enable Plugin');

    btn.simulate('click');

    setTimeout(() => {
      try {
        component.update();
        expect(btn.text()).toBe('Disable Plugin');
        done();
      } catch (err) {
        done(err);
      }
    }, 1);
  });
});
