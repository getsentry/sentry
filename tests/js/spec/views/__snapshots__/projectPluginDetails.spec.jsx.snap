// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectPluginDetails renders 1`] = `
.glamor-4 {
  font-size: 14px;
  box-shadow: inset 0 -1px 0;
  margin: -20px 0 30px;
}

.glamor-2 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.glamor-0 {
  font-size: 20px;
  font-weight: bold;
  margin: 20px 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<withPlugins
  location={
    Object {
      "pathame": "/mock-pathname/",
      "query": Object {},
    }
  }
  organization={
    Object {
      "access": Array [
        "org:read",
        "org:write",
        "org:admin",
        "project:read",
        "project:write",
        "project:admin",
        "team:read",
        "team:write",
        "team:admin",
      ],
      "features": Array [],
      "id": "3",
      "name": "Organization Name",
      "onboardingTasks": Array [],
      "slug": "org-slug",
      "status": Object {
        "id": "active",
        "name": "active",
      },
      "teams": Array [],
    }
  }
  params={
    Object {
      "orgId": "org-slug",
      "pluginId": "amazon-sqs",
      "projectId": "project-slug",
    }
  }
  project={
    Object {
      "digestsMaxDelay": 60,
      "digestsMinDelay": 5,
      "id": "2",
      "name": "Project Name",
      "slug": "project-slug",
      "subjectTemplate": "[$project] \${tag:level}: $title",
    }
  }
>
  <ProjectPluginDetails
    location={
      Object {
        "pathame": "/mock-pathname/",
        "query": Object {},
      }
    }
    organization={
      Object {
        "access": Array [
          "org:read",
          "org:write",
          "org:admin",
          "project:read",
          "project:write",
          "project:admin",
          "team:read",
          "team:write",
          "team:admin",
        ],
        "features": Array [],
        "id": "3",
        "name": "Organization Name",
        "onboardingTasks": Array [],
        "slug": "org-slug",
        "status": Object {
          "id": "active",
          "name": "active",
        },
        "teams": Array [],
      }
    }
    params={
      Object {
        "orgId": "org-slug",
        "pluginId": "amazon-sqs",
        "projectId": "project-slug",
      }
    }
    plugins={
      Object {
        "error": null,
        "loading": true,
        "pageLinks": null,
        "plugins": Array [],
      }
    }
    project={
      Object {
        "digestsMaxDelay": 60,
        "digestsMinDelay": 5,
        "id": "2",
        "name": "Project Name",
        "slug": "project-slug",
        "subjectTemplate": "[$project] \${tag:level}: $title",
      }
    }
  >
    <SideEffect(DocumentTitle)
      title="Sentry"
    >
      <DocumentTitle
        title="Sentry"
      >
        <div>
          <SettingsPageHeading
            action={
              <div
                className="pull-right"
              >
                <Button
                  disabled={false}
                  onClick={[Function]}
                  style={
                    Object {
                      "marginRight": "6px",
                    }
                  }
                >
                  Enable Plugin
                </Button>
                <Button
                  disabled={false}
                  onClick={[Function]}
                >
                  Reset Configuration
                </Button>
              </div>
            }
            title="Amazon SQS"
          >
            <Styled(div)>
              <div
                className="glamor-4 glamor-5"
              >
                <Flex
                  align="center"
                >
                  <Base
                    align="center"
                    className="glamor-2"
                  >
                    <div
                      className="glamor-2"
                      is={null}
                    >
                      <Styled(div)>
                        <div
                          className="glamor-0 glamor-1"
                        >
                          Amazon SQS
                        </div>
                      </Styled(div)>
                      <div>
                        <div
                          className="pull-right"
                        >
                          <Button
                            disabled={false}
                            onClick={[Function]}
                            style={
                              Object {
                                "marginRight": "6px",
                              }
                            }
                          >
                            <button
                              className="button button-default"
                              disabled={false}
                              onClick={[Function]}
                              role="button"
                              style={
                                Object {
                                  "marginRight": "6px",
                                }
                              }
                            >
                              <FlowLayout
                                truncate={false}
                              >
                                <div
                                  className="flow-layout"
                                >
                                  <span
                                    className="button-label"
                                  >
                                    Enable Plugin
                                  </span>
                                </div>
                              </FlowLayout>
                            </button>
                          </Button>
                          <Button
                            disabled={false}
                            onClick={[Function]}
                          >
                            <button
                              className="button button-default"
                              disabled={false}
                              onClick={[Function]}
                              role="button"
                            >
                              <FlowLayout
                                truncate={false}
                              >
                                <div
                                  className="flow-layout"
                                >
                                  <span
                                    className="button-label"
                                  >
                                    Reset Configuration
                                  </span>
                                </div>
                              </FlowLayout>
                            </button>
                          </Button>
                        </div>
                      </div>
                    </div>
                  </Base>
                </Flex>
              </div>
            </Styled(div)>
          </SettingsPageHeading>
          <div
            className="row"
          >
            <div
              className="col-md-7"
            >
              <PluginConfig
                data={
                  Object {
                    "assets": Array [],
                    "author": Object {
                      "name": "Sentry Team",
                      "url": "https://github.com/getsentry/sentry",
                    },
                    "canDisable": true,
                    "enabled": false,
                    "hasConfiguration": true,
                    "id": "amazon-sqs",
                    "name": "Amazon SQS",
                    "slug": "amazon-sqs",
                    "version": "8.23.0.dev0",
                  }
                }
                enabled={false}
                onDisablePlugin={[Function]}
                organization={
                  Object {
                    "access": Array [
                      "org:read",
                      "org:write",
                      "org:admin",
                      "project:read",
                      "project:write",
                      "project:admin",
                      "team:read",
                      "team:write",
                      "team:admin",
                    ],
                    "features": Array [],
                    "id": "3",
                    "name": "Organization Name",
                    "onboardingTasks": Array [],
                    "slug": "org-slug",
                    "status": Object {
                      "id": "active",
                      "name": "active",
                    },
                    "teams": Array [],
                  }
                }
                project={
                  Object {
                    "digestsMaxDelay": 60,
                    "digestsMinDelay": 5,
                    "id": "2",
                    "name": "Project Name",
                    "slug": "project-slug",
                    "subjectTemplate": "[$project] \${tag:level}: $title",
                  }
                }
              >
                <div
                  className="box ref-plugin-config-amazon-sqs"
                >
                  <div
                    className="box-header"
                  >
                    <h3>
                      Amazon SQS
                    </h3>
                  </div>
                  <div
                    className="box-content with-padding"
                  >
                    <div
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": undefined,
                        }
                      }
                    />
                    <PluginSettings
                      organization={
                        Object {
                          "access": Array [
                            "org:read",
                            "org:write",
                            "org:admin",
                            "project:read",
                            "project:write",
                            "project:admin",
                            "team:read",
                            "team:write",
                            "team:admin",
                          ],
                          "features": Array [],
                          "id": "3",
                          "name": "Organization Name",
                          "onboardingTasks": Array [],
                          "slug": "org-slug",
                          "status": Object {
                            "id": "active",
                            "name": "active",
                          },
                          "teams": Array [],
                        }
                      }
                      plugin={
                        DefaultPlugin {
                          "assets": Array [],
                          "author": Object {
                            "name": "Sentry Team",
                            "url": "https://github.com/getsentry/sentry",
                          },
                          "canDisable": true,
                          "enabled": false,
                          "hasConfiguration": true,
                          "id": "amazon-sqs",
                          "name": "Amazon SQS",
                          "slug": "amazon-sqs",
                          "version": "8.23.0.dev0",
                        }
                      }
                      project={
                        Object {
                          "digestsMaxDelay": 60,
                          "digestsMinDelay": 5,
                          "id": "2",
                          "name": "Project Name",
                          "slug": "project-slug",
                          "subjectTemplate": "[$project] \${tag:level}: $title",
                        }
                      }
                    />
                  </div>
                </div>
              </PluginConfig>
            </div>
            <div
              className="col-md-4 col-md-offset-1"
            >
              <div
                className="pluginDetails-meta"
              >
                <h4>
                  Plugin Information
                </h4>
                <dl
                  className="flat"
                >
                  <dt>
                    Name
                  </dt>
                  <dd>
                    Amazon SQS
                  </dd>
                  <dt>
                    Author
                  </dt>
                  <dd>
                    Sentry Team
                  </dd>
                  <div>
                    <dt>
                      URL
                    </dt>
                    <dd>
                      <ExternalLink
                        href="https://github.com/getsentry/sentry"
                        rel="noreferrer noopener"
                        target="_blank"
                      >
                        <a
                          href="https://github.com/getsentry/sentry"
                          rel="noreferrer noopener"
                          target="_blank"
                        >
                          github.com/getsentry/sentry
                        </a>
                      </ExternalLink>
                    </dd>
                  </div>
                  <dt>
                    Version
                  </dt>
                  <dd>
                    8.23.0.dev0
                  </dd>
                </dl>
              </div>
            </div>
          </div>
        </div>
      </DocumentTitle>
    </SideEffect(DocumentTitle)>
  </ProjectPluginDetails>
</withPlugins>
`;
